### PR TITLE
feat(logging): add initial supporting structures for new logger plugin

### PR DESCRIPTION
**Pull request by phantumcode**
## General Checklist

- [x] Added new tests to cover change, if needed
- [x] Build succeeds with all target using Swift Package Manager
- [x] All unit tests pass
- [x] All integration tests pass
- [x] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)
- [x] Documentation update for the change if required
- [x] PR title conforms to conventional commit style
- [x] New or updated tests include `Given When Then` inline code documentation and are named accordingly `testThing_condition_expectation()`
- [x] If breaking change, documentation/changelog update with migration instructions

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Amplify/DefaultPlugins/AWSUnifiedLoggingPlugin/Internal/OSLogWrapper.swift [hunk 1]
```diff
@@ -9,7 +9,9 @@ import os.log
 
 final class OSLogWrapper: Logger {
     private let osLog: OSLog
-
+    
+    var enable: Bool = true
```
Comment (thisisabhash) [inline Amplify/DefaultPlugins/AWSUnifiedLoggingPlugin/Internal/OSLogWrapper.swift:13]: nit: rename to `enabled` or `loggingEnabled`

File: AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/Persistence/LogEntry.swift [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+struct LogEntry: Codable, Hashable {
+    
+    /// The timestamp representing the creation time of the log entry or event.
+    let created: Date
+
+    /// The logical tag of the log entry or event. This will likely be a
+    /// Category name.
+    let category: String
+    
+    let namespace: String?
```
Comment (thisisabhash) [inline AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/Persistence/LogEntry.swift:21]: nit: can we have a comment explaining this property

File: AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/Persistence/LogEntryCodec.swift [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            throw DecodingError.stringNotUtf8(string)
+        }
+        return try decode(data: data)
+    }
+    
+    func decode(from fileURL: URL) throws -> [LogEntry] {
+        guard let scheme = fileURL.scheme else {
+            throw DecodingError.invalidScheme(log: fileURL)
+        }
+        guard scheme == "file" else {
```
Comment (thisisabhash) [inline AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/Persistence/LogEntryCodec.swift:47]: can we use this utility method? 
https://developer.apple.com/documentation/foundation/url/1780396-isfileurl

File: AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/Persistence/LogEntryCodec.swift [hunk 4]
```diff
@@ -0,0 +1,60 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+struct LogEntryCodec {
```
Comment (thisisabhash) [inline AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/Persistence/LogEntryCodec.swift:10]: documentation: Can we add comments explaining this `struct`

File: AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/Producer/RotatingLogBatch.swift [hunk 5]
```diff
@@ -0,0 +1,34 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+struct RotatingLogBatch {
```
Comment (thisisabhash) [inline AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/Producer/RotatingLogBatch.swift:10]: documentation: Can we add comments explaining this `struct`

File: AmplifyPlugins/Logging/Tests/AWSCloudWatchLoggingPluginTests/CloudWatchLogConsumerTests.swift [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+        entries = nil
+        interactions = nil
+        client = nil
+        logGroupName = nil
+        logStreamName = nil
+        systemUnderTest = nil
+    }
+    
+    /// - Given: A list of a single log entry
+    /// - When: CloudWatchLoggingConsumer consumes the a log batch
```
Comment (thisisabhash) [inline AmplifyPlugins/Logging/Tests/AWSCloudWatchLoggingPluginTests/CloudWatchLogConsumerTests.swift:43]: nit typo: the
Similar in tests below

File: AmplifyPlugins/Logging/Tests/AWSCloudWatchLoggingPluginTests/LogEntryTests.swift [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    /// Given: a Log Entry json with invalid log level
+    /// When: decoding occurs
+    /// Then: the log level defaults to Error
+    func testDecodeDefaultsLogLevelWithInvalidLogLevel() throws {
+        let message = UUID().uuidString
+        let json = """
+        {
+            "created": 0,
+            "category": "LogEntryTests",
+            "level": -1337,
```
Comment (thisisabhash) [inline AmplifyPlugins/Logging/Tests/AWSCloudWatchLoggingPluginTests/LogEntryTests.swift:62]: QQ: what is the -1337 value?

File: Amplify/DefaultPlugins/AWSUnifiedLoggingPlugin/Internal/OSLogWrapper.swift (hunk 1 above)
Comment (phantumcode) [reply Amplify/DefaultPlugins/AWSUnifiedLoggingPlugin/Internal/OSLogWrapper.swift]: variable renamed to `enabled` as suggested

File: AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/Persistence/LogEntry.swift (hunk 2 above)
Comment (phantumcode) [reply AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/Persistence/LogEntry.swift:23]: missing comment added

File: AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/Persistence/LogEntryCodec.swift (hunk 3 above)
Comment (phantumcode) [reply AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/Persistence/LogEntryCodec.swift]: updated to use utility method `isFileUrl`

File: AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/Persistence/LogEntryCodec.swift (hunk 4 above)
Comment (phantumcode) [reply AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/Persistence/LogEntryCodec.swift:11]: added missing comments

File: AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/Producer/RotatingLogBatch.swift (hunk 5 above)
Comment (phantumcode) [reply AmplifyPlugins/Logging/Sources/AWSCloudWatchLoggingPlugin/Producer/RotatingLogBatch.swift:12]: added missing comments

File: AmplifyPlugins/Logging/Tests/AWSCloudWatchLoggingPluginTests/CloudWatchLogConsumerTests.swift (hunk 6 above)
Comment (phantumcode) [reply AmplifyPlugins/Logging/Tests/AWSCloudWatchLoggingPluginTests/CloudWatchLogConsumerTests.swift]: fixed typo

File: AmplifyPlugins/Logging/Tests/AWSCloudWatchLoggingPluginTests/LogEntryTests.swift (hunk 7 above)
Comment (phantumcode) [reply AmplifyPlugins/Logging/Tests/AWSCloudWatchLoggingPluginTests/LogEntryTests.swift]: it's meant to be any invalid value below 0, updated the comments with clarification

**Review by thisisabhash — APPROVED** [review]
LGTM

**Pull request merged**